src/app/main.rs

```diff
@@ -54,9 +54,10 @@ use super::profiles::Client as ProfilesClient;
 ///
 /// The private listener routes requests to service-discovery-aware load-balancer.
 ///
-pub struct Main<G> {
+pub struct Main<G, R = dns::DefaultResolver> {
     proxy_parts: ProxyParts<G>,
     runtime: task::MainRuntime,
+    dns: R
```

Comment:
We have three different names going on for DNS resolvers. Here we're calling it `dns`, but we call it `dns_resolver` and `resolver` in other parts of the proxy. Maybe we could consolidate down one name? Otherwise looks good!